### PR TITLE
Add JIT-helper of ArgTypeNotSupportedException

### PR DESCRIPTION
**Pull request by fiigii**
Intel hardware intrinsics need to throw exceptions with `System.NotSupportedException: Specified type is not supported` on `Vector128/256<T>` that is instantiated by non-numeric types.

### CODE REVIEW

**Comment by fiigii** [comment]
@CarolEidt @jkotas Would you like to port this change to desktop CLR at first? I need this jit-helper in my `Vector128/256<T>` work.

**Comment by jkotas** [comment]
You will need to change JIT interface GUID at the top of corinfo..h for this change.

cc @dotnet/jit-contrib

File: src/inc/corinfo.h [hunk 1]
```diff
@@ -643,6 +643,7 @@ enum CorInfoHelpFunc
     CORINFO_HELP_THROW_ARGUMENTEXCEPTION,           // throw ArgumentException
     CORINFO_HELP_THROW_ARGUMENTOUTOFRANGEEXCEPTION, // throw ArgumentOutOfRangeException
     CORINFO_HELP_THROW_PLATFORM_NOT_SUPPORTED,      // throw PlatformNotSupportedException
+    CORINFO_HELP_THROW_ARGTYPE_NOT_SUPPORTED,               // throw NotSupportedException
```
Comment (janvorli) [inline src/inc/corinfo.h]: A nit - comment alignment

Comment (fiigii) [reply src/inc/corinfo.h]: Thank you, will fix.

**Comment by fiigii** [comment]
> You will need to change JIT interface GUID at the top of corinfo..h for this change.

@jkotas I see. Shall I update GUID now? Or wait for the update from desktop CLR change?

**Comment by briansull** [comment]
Go ahead and make the GUID change as part of your checkin.

We will need to check in the `src/inc/corinfo.h` changes on the Desktop as a separate step.

**Comment by fiigii** [comment]
@briansull Got it, thank you.

**Comment by fiigii** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test
@dotnet-bot test Tizen armel Cross Checked Innerloop Build and Test 
@dotnet-bot test Ubuntu x64 Checked Innerloop Build and Test 
@dotnet-bot test Ubuntu16.04 armlb Cross Debug Innerloop Build

**Comment by fiigii** [comment]
@jkotas The test failures look unrelated to this change?

**Comment by jkotas** [comment]
Actually, I do not think you need a helper for this. Why can't the jit fallback to the non-inlined implementation to throw the exception?

**Comment by fiigii** [comment]
> Why can't the jit fallback to the non-inlined implementation to throw the exception?

@jkotas Do you mean something like this?
```csharp
public static Vector128<T> GetLowerHalf<T>(Vector256<T> value) where T : struct 
{
    if (typeof(T) != typeof(Byte) && typeof(T) != typeof(UByte) && ... )
    {
        throw new NotSupportedException(SR.Arg_TypeNotSupported);
    }
    return GetLowerHalf<T>(value);
}
```
This is a good point. However, our current contract is "named intrinsics that **directly called** by itself must be expanded". If we add the exception code in the managed implementation, that would break this contract. Then other named intrinsics cannot have the fallback with any recursive algorithm in the future. I am afraid that might be a too strict limitation.

**Comment by jkotas** [comment]
Right, something like this.

I would think that `gtIsRecursiveCall` will still return true for the above case and the inner call will be force expanded. If there is something that I am missing, we should fix it.

**Comment by fiigii** [comment]
> I would think that gtIsRecursiveCall will still return true for the above case and the inner call will be force expanded. 

I meant maybe other recursive intrinsics (in the future) that we still want them optionally expanded and only "directly recursive" intrinsics are forced expanded. 
But, at this time, your solution is reasonable to me. If you think we do not need to consider the above situation, I will close this PR.

**Comment by jkotas** [comment]
Right, I do not think we need to worry about it - until we have a real situation like that.

File: src/vm/jithelpers.cpp [hunk 2]
```diff
@@ -5127,6 +5127,24 @@ HCIMPL0(void, JIT_ThrowPlatformNotSupportedException)
...
+    FCALL_CONTRACT;
+
+    /* Make no assumptions about the current machine state */
+    ResetCurrentContext();
+
+    FC_GC_POLL_NOT_NEEDED();    // throws always open up for GC
+
+    HELPER_METHOD_FRAME_BEGIN_ATTRIB_NOPOLL(Frame::FRAME_ATTR_EXCEPTION);    // Set up a frame
+
+    COMPlusThrow(kNotSupportedException, W("Arg_TypeNotSupported"));
```
Comment (fiigii) [inline src/vm/jithelpers.cpp]: This behavior is slightly different from `S.N.Vector<T>` that throws NSE with message "Specified type is not supported".

https://github.com/dotnet/coreclr/pull/15028

**Pull request closed without merging**